File: cplusplus/src/create_calculate_safety_number.cpp
#include "autograph/create_calculate_safety_number.h"

#include <algorithm>
#include <string>

#include "autograph/calculate_safety_number.h"
#include "autograph/constants.h"

CalculateSafetyNumberFunction create_calculate_safety_number(
    const Chunk &our_identity_key) {
  auto calculate_safety_number_function =
      [&our_identity_key](const Chunk &their_identity_key) {
        Chunk safety_number(SAFETY_NUMBER_SIZE);
        bool success = calculate_safety_number(safety_number.data(),
                                               our_identity_key.data(),
                                               their_identity_key.data());
        if (!success) {
          throw std::runtime_error("Failed to calculate safety number");
        }
        return std::move(safety_number);
      };
  return std::move(calculate_safety_number_function);
}

File: typescript/src/create-calculate-safety-number.ts
import { Chunk, createFrom } from 'stedy/bytes'
import { CalculateSafetyNumberFunction } from '../types'
import { SAFETY_NUMBER_DIVISOR, SAFETY_NUMBER_ITERATIONS } from './constants'
import hash from './crypto/hash'

const encodeChunk = (chunk: Chunk) => {
  const [a, b, c, d, e] = chunk
  const number =
    (a * 2 ** 32 + b * 2 ** 24 + c * 2 ** 16 + d * 2 ** 8 + e) %
    SAFETY_NUMBER_DIVISOR
  const result = number.toString()
  return `${'0'.repeat(5 - result.length)}${result}`
}

const calculate = async (identityKey: BufferSource) => {
  const digest = await hash(identityKey, SAFETY_NUMBER_ITERATIONS)
  return digest
    .subarray(0, 30)
    .split(5)
    .map((chunk) => encodeChunk(chunk))
    .join('')
}

const calculateSafetyNumber = async (
  ourIdentityKey: BufferSource,
  theirIdentityKey: BufferSource
) => {
  const fingerprints = await Promise.all([
    calculate(ourIdentityKey),
    calculate(theirIdentityKey)
  ])
  return createFrom(fingerprints.sort().join(''))
}

const createCalculateSafetyNumber =
  (ourIdentityKey: BufferSource): CalculateSafetyNumberFunction =>
  (theirIdentityKey: BufferSource) =>
    calculateSafetyNumber(ourIdentityKey, theirIdentityKey)

export default createCalculateSafetyNumber
